bindgen/TypeTranslator.cpp

```diff
@@ -68,111 +57,143 @@ TypeTranslator::TranslateFunctionPointer(const clang::QualType &qtpe,
 
         // Take care of void*
         if (as->getKind() == clang::BuiltinType::Void) {
-            return "native.Ptr[Byte]";
+            return new PointerType(new PrimitiveType("Byte"));
         }
 
         // Take care of char*
         if (as->getKind() == clang::BuiltinType::Char_S ||
             as->getKind() == clang::BuiltinType::SChar) {
-            return "native.CString";
+            return new PrimitiveType("native.CString");
         }
     }
 
-    return std::string("native.Ptr[") + Translate(pte, avoid) +
-           std::string("]");
+    return new PointerType(translate(pte, avoid));
 }
 
-std::string
-TypeTranslator::TranslateStructOrUnion(const clang::QualType &qtpe) {
+Type *TypeTranslator::translateStruct(const clang::QualType &qtpe) {
     if (qtpe->hasUnnamedOrLocalType()) {
         // TODO: Verify that the local part is not a problem
         uint64_t size = ctx->getTypeSize(qtpe);
-        return "native.CArray[Byte, " + uint64ToScalaNat(size) + "]";
+        return new ArrayType(new PrimitiveType("Byte"), size);
     }
 
     std::string name = qtpe.getUnqualifiedType().getAsString();
 
     // TODO: do it properly
     size_t f = name.find(std::string("struct __dirstream"));
     if (f != std::string::npos) {
-        return std::string("native.CArray[Byte, Digit[_3, Digit[_2, _0]]]");
+        return new ArrayType(new PrimitiveType("Byte"), 320);
     }
 
-    f = name.find(" ");
-    if (f != std::string::npos) {
-        return name.replace(f, std::string(" ").length(), "_");
+    auto it = aliasesMap.find(name);
+    if (it != aliasesMap.end()) {
+        /* name contains space: struct <name>.
+         * Use type alias instead struct type */
+        return (*it).second;
     }
-    return name;
+    /* type has typedef alias */
+    return ir.getTypeDefWithName(name);
 }
 
-std::string TypeTranslator::TranslateEnum(const clang::QualType &qtpe) {
+Type *TypeTranslator::translateUnion(const clang::QualType &qtpe) {
+    if (qtpe->hasUnnamedOrLocalType()) {
+        // TODO: Verify that the local part is not a problem
+        uint64_t size = ctx->getTypeSize(qtpe);
+        return new ArrayType(new PrimitiveType("Byte"), size);
+    }
+
     std::string name = qtpe.getUnqualifiedType().getAsString();
-    size_t f = name.find(" ");
-    if (f != std::string::npos) {
-        return name.replace(f, std::string(" ").length(), "_");
+
+    auto it = aliasesMap.find(name);
+    if (it != aliasesMap.end()) {
+        /* name contains space: union <name>.
+         * Use type alias instead union type */
+        return (*it).second;
     }
-    return name;
+    /* type has typedef alias */
+    return ir.getTypeDefWithName(name);
 }
 
-std::string
-TypeTranslator::TranslateConstantArray(const clang::ConstantArrayType *ar,
-                                       const std::string *avoid) {
+Type *TypeTranslator::translateEnum(const clang::QualType &qtpe) {
```

Comment:
It looks like this method is reused actually be used by `translateStruct` or even merged with it once the `union` specific `if` branch is removed.

Reply:
:+1: